File: workers/worker-php/src/worker_runner.php
<?php

require __DIR__ . '/../vendor/autoload.php';

use OpenRegex\Worker\Registry;
use OpenRegex\Worker\Processor;

$redisUrl = getenv('REDIS_URL') ?: 'redis://redis:6379';
$client = new Predis\Client($redisUrl);

// Graceful unregister on shutdown (docker stop / Ctrl+C).
if (function_exists('pcntl_async_signals')) {
    pcntl_async_signals(true);
    $unregister = function () use ($client) {
        try {
            $client->hdel('openregex:workers', ['worker-php']);
            $client->del(['openregex:workers:heartbeat:worker-php']);
        } catch (\Exception $e) {
            // best effort
        }
        exit(0);
    };
    pcntl_signal(SIGTERM, $unregister);
    pcntl_signal(SIGINT, $unregister);
}

try {
    // Heartbeat must be live before registration so discovery never sees
    // a registered worker without one.
    $client->setex('openregex:workers:heartbeat:worker-php', 15, (string)time());
    Registry::registerEngines($client);
    Processor::listenAndProcess($client);
} catch (\Exception $e) {
    echo "Fatal error during startup: " . $e->getMessage() . "\n";
    exit(1);
}

File: workers/worker-jvm8/src/main/java/com/openregex/jvm8/WorkerRunner.java
package com.openregex.jvm8;

import redis.clients.jedis.Jedis;

public final class WorkerRunner {

    private static final String WORKERS_HASH_KEY = "openregex:workers";
    private static final String HEARTBEAT_KEY = "openregex:workers:heartbeat:" + Registry.WORKER_NAME;
    private static final int HEARTBEAT_TTL_S = 15;
    private static final long HEARTBEAT_INTERVAL_MS = 5000L;

    private WorkerRunner() {
    }

    private static void setHeartbeat(Jedis jedis) {
        try {
            jedis.setex(HEARTBEAT_KEY, HEARTBEAT_TTL_S, String.valueOf(System.currentTimeMillis() / 1000L));
        } catch (Exception ignored) {
            // transient Redis outage; the TTL just expires until it recovers
        }
    }

    public static void main(String[] args) {
        String redisUrl = System.getenv("REDIS_URL");
        if (redisUrl == null || redisUrl.isEmpty()) {
            redisUrl = "redis://redis:6379";
        }
        final String finalRedisUrl = redisUrl;

        // Heartbeat must be live before registration so discovery never sees
        // a registered worker without one.
        Jedis registrar = new Jedis(finalRedisUrl);
        try {
            setHeartbeat(registrar);
            Registry.registerEngines(registrar);
        } catch (Exception e) {
            System.err.println("Failed to register engines: " + e.getMessage());
        } finally {
            registrar.close();
        }

        Thread heartbeat = new Thread(new Runnable() {
            public void run() {
                Jedis beat = new Jedis(finalRedisUrl);
                while (!Thread.currentThread().isInterrupted()) {
                    try {
                        Thread.sleep(HEARTBEAT_INTERVAL_MS);
                    } catch (InterruptedException e) {
                        Thread.currentThread().interrupt();
                        return;
                    }
                    setHeartbeat(beat);
                }
            }
        }, "worker-heartbeat");
        heartbeat.setDaemon(true);
        heartbeat.start();

        // Graceful unregister on shutdown (docker stop / Ctrl+C).
        Runtime.getRuntime().addShutdownHook(new Thread(new Runnable() {
            public void run() {
                Jedis jedis = new Jedis(finalRedisUrl);
                try {
                    jedis.hdel(WORKERS_HASH_KEY, Registry.WORKER_NAME);
                    jedis.del(HEARTBEAT_KEY);
                } catch (Exception ignored) {
                    // best effort
                } finally {
                    jedis.close();
                }
            }
        }, "worker-unregister"));

        Processor.listenAndProcess(finalRedisUrl);
    }
}
